stdlib/src/collections/deque.mojo

```diff
@@ -603,6 +625,259 @@ struct Deque[ElementType: CollectionElement](
             self._head = self._physical_index(self._head - 1)
             (src + i).move_pointee_into(self._data + self._head)
 
+    fn index[
+        EqualityElementType: EqualityComparableCollectionElement, //
+    ](
+        self: Deque[EqualityElementType],
+        value: EqualityElementType,
+        start: Int = 0,
+        stop: Optional[Int] = None,
+    ) raises -> Int:
+        """Returns the index of the first occurrence of a `value` in a deque
+        restricted by the range given the `start` and `stop` bounds.
+
+        Parameters:
+            EqualityElementType: The type of the elements in the deque.
+                Must implement the `EqualityComparableCollectionElement` trait.
+
+        Args:
+            value: The value to search for.
+            start: The starting index of the search, treated as a slice index
+                (defaults to 0).
+            stop: The ending index of the search, treated as a slice index
+                (defaults to None, which means the end of the deque).
+
+        Returns:
+            The index of the first occurrence of the value in the deque.
+
+        Raises:
+            ValueError: If the value is not found in the deque.
+        """
+        start_normalized = start
+
+        if stop is None:
+            stop_normalized = len(self)
+        else:
+            stop_normalized = stop.value()
+
+        if start_normalized < 0:
+            start_normalized += len(self)
+        if stop_normalized < 0:
+            stop_normalized += len(self)
+
+        start_normalized = max(0, min(start_normalized, len(self)))
+        stop_normalized = max(0, min(stop_normalized, len(self)))
+
+        for idx in range(start_normalized, stop_normalized):
+            offset = self._physical_index(self._head + idx)
+            if (self._data + offset)[] == value:
+                return idx
+        raise "ValueError: Given element is not in deque"
```

Comment:
That would restrict `index()` to be used only on the deques with the StringRepresentable element types. I think this would be an unnecessary restriction, especially for the user defined types.

Reply:
I agree re the unnecessary restriction, thanks for the reminder on that 👍